Marlin/src/module/stepper.cpp

```diff
@@ -2462,7 +2468,7 @@ hal_timer_t Stepper::block_phase_isr() {
        */
       #if ENABLED(LASER_POWER_TRAP)
         if (cutter.cutter_mode == CUTTER_MODE_CONTINUOUS) {
-          if (step_events_completed + 1 == accelerate_until) {
+          if (step_events_completed + 1 == accelerate_before) {
```

Comment:
Well, the new naming did highlight this: I think it should be without the +1 but I'm not sure.

Reply:
We'll have to analyze the intent of this line and look at its history. Previously it wanted to match `accelerate_until - 1` which would be the second-to-last step of acceleration. Assuming that we want no change in behavior, and assuming that `accelerate_before` is now one more than `accelerate_until`, the actual correction would be `step_events_completed + 2 == accelerate_before`.

Again, we need to look closer at the original intent of the line, because, for one thing, `step_events_completed` is not incremented by 1, but by `events_to_do` which may or may not always line up correctly to match the exact index `accelerate_before - n`. Once we fully grok the intent of this line we can simply replace it with whatever best fits the updated code.

Reply:
Let's see if @descipher has anything to add….

Reply:
Well it's been quite some time since my head deep inside this code! "first block->accelerate_until step_events_completed" is the key element, since the first block starts at 0 we will have value 1 after the ISR does the count increment, thus the <  accelerate_until naming usage which is not the same meaning as accelerate_before, it only looks that way when viewing it as less than accelerate_until at the moment of comparison. However this comparison moment is the actual accelerate_until PIT and not before.  

In regards to the +1 == if I recall correctly the reason for + 1 stems from an odd step count and the PIT when the laser would need to be turned off or incr/decr . It worked out that the counts were shifted toward the tail of the trap so the counts were consistently short on the decel side of the trap. This can be viewed by turning on debug where it will give you the actual step counts. It becomes obvious with debug on and the simulator is your friend for that ;) The laser power will be inc'd and dec'd closer to ideal with the +1. Keep in mind its not uniform in a real world print or laser burn, but in tests from a dead start to full stop the ramps should be equal and that's when this shifted count behavior was first observed. If there is a change in the counts/shaping now it should be re-evaluated again.

Reply:
My worry is that with some moves it will never be that `step_events_completed + 1 == accelerate_before` during the `block_phase_isr` because AFAIK this code is not guaranteed to be executed after every step, but only when some new "block phase" is "needed."

The best thing to do would be to test this in the simulator, not with laser enabled, but just with that comparison being done at that same point in the code to see how it matches up and whether we are in fact guaranteed to be able to scrutinize every step index. If we can, then the very last index of acceleration will be `accelerate_before - 1` and we can apply that correction.

Reply:
> The best thing to do would be to test this in the simulator

As a reminder, it is possible to log signals in the simulator and export them to files that can be viewed in PulseView. This allows  you to analyze stepper pulses without requiring a physical logic analyzer setup. I find that it provides a much faster feedback loop than working with actual hardware.

It won't work for hardware PWM signals like you would have for a laser, but for any software-timed pulses it works well.

Reply:
PR Sim compiling result:
Took a look at the simulation outputs and captured a logic vcd of (X|Y)_STEP_PIN it looks reasonable however there is still the count shifting towards the tail of the trap where we can observe the resulting faster decel step rate. 
This is the X_STEP_PIN Y_STEP_PIN output point and the zoom is the X axis. 
![image](https://github.com/MarlinFirmware/Marlin/assets/10734527/9e68fbbb-2ab8-4567-9498-e9f05f66d99f)
![image](https://github.com/MarlinFirmware/Marlin/assets/10734527/7d9cfcfc-3a30-47d0-b0f6-817d2607cdb1)
Here the start pulse interval is 3.187ms and the stop is 2.27ms. This is a dead stop to dead stop thus we should have equal sides on the trap. We have a significantly harder decel due to the shorter decel count calc.
@thinkyhead We can use ( + 1 >=) to ensure we set the laser power on a missed compare. The intent of this code is to set the power to the current_block planner calculated "Cruise" laser power value. Since this code only happens on the cruise else section it should have no negative impact with the addition of >. If we end up fixing the counts then the +1 can be removed.

Reply:
Since there are `LASER_POWER_TRAP` specific issues I've created #27031 to address them. The issues addressed in a way that's independent of this change.

Now to address how this change affects laser: I looked over how `LASER_POWER_TRAP` uses `ac/decelerate_steps` and (with the #27031 fixes applied too) the resulting ramp up/down is smoother than before. Here's a comparison at the end of a deceleration when the laser is going to 0% power.

Before:
![laser-before-dec](https://github.com/MarlinFirmware/Marlin/assets/299015/e364ae7d-76b5-465f-a8ef-dc41afd043ff)
(just a note that my steppers are set to edge stepping)

After:
![laser-smooth-dec](https://github.com/MarlinFirmware/Marlin/assets/299015/e80e372d-37aa-464a-a1b2-b518ac3bc275)

I've also looked at the acceleration->cruise transition. I'd say the difference is not significant / not above the noise in the pwm/logic analyzer capture: Last step 75.4 -> 80% versus 76.7% -> 80%.

> Took a look at the simulation outputs and captured a logic vcd of (X|Y)_STEP_PIN it looks reasonable however there is still the count shifting towards the tail of the trap where we can observe the resulting faster decel step rate. [..]
> Here the start pulse interval is 3.187ms and the stop is 2.27ms. This is a dead stop to dead stop thus we should have equal sides on the trap. We have a significantly harder decel due to the shorter decel count calc.

@descipher You are right about the absolute ideal being equal intervals. And that even with this change the first step is slower than the last step, or stated another way: the first step is at exactly `initial_rate` but the last step isn't yet at `final_rate`. Two answers:
- the difference was way worse before this change as the decel count was even lower
- trying to hit the absolute ideal profile results in a code complexity explosion. This is because the `ac/deceleration_steps` computation has to work for all the following types of segments: full acceleration, full deceleration, acceleration->deceleration, acceleration->cruise->deceleration (5 cases). And say we do that, and get the first step at exactly `initial_rate` and the last at exactly `final_rate` (=`initial_rate` of next segment). Now the full acceleration->full acceleration and the full deceleration->full deceleration transitions become jerky because when `final_rate` and `initial_rate` of the next segment are equal that means there's no acceleration happening between segments, i.e. we go full accel, 0 accel, full accel. Special casing the 5 cases into 5 cases transitions is just too much. Also the float->discrete steps forces us to accept an error somewhere. To avoid this complexity explosion the way this code is structured is to accept an error of ~1/2 of the ideal acceleration. This half error manifests as not hitting the `final_rate` exactly.